### ISSUE TITLE
Cache multi-stage builds

### ISSUE THREAD

**Issue by chrisjsewell** [elgohr/Publish-Docker-Github-Action#87]
For example with the following, the first `RUN` never uses the cache.

```dockerfile
FROM busybox as builder
RUN echo "hello" > test

FROM busybox
COPY --from=builder test test
RUN echo test
```

**Comment by elgohr** [comment]
This can be achieved by using buildkit (https://github.com/moby/moby/issues/34715). As buildkit is only able to build Linux containers (https://docs.docker.com/develop/develop-images/build_enhancements/#limitations), but GitHub actions also supports other platforms, I'm not sure whether it should be supported in this action.

By the way, there's a workaround https://stackoverflow.com/a/52649913

**Comment by chrisjsewell** [comment]
Cheers for the response, sorry I'm a bit busy at the moment, but I'll try to test this in due course

**Comment by linkdd** [comment]
Nope, even with the ``DOCKER_BUILDKIT`` environment variable, the cache is still not being used.

**Comment by elgohr** [comment]
Maybe you want to cache it from a previous build?
https://github.com/elgohr/Publish-Docker-Github-Action#cache

**Comment by linkdd** [comment]
Since it's in the README, that is the first option I tried.

Here is the complete configuration I'm using:

```yaml
      - name: myimage@docker
        uses: elgohr/Publish-Docker-Github-Action@master
        env:
          DOCKER_BUILDKIT: "1"
        with:
          name: myorga/myrepo/myimage
          username: ${{ secrets.DOCKER_USERNAME }}
          password: ${{ secrets.DOCKER_PASSWORD }}
          registry: docker.pkg.github.com
          dockerfile: containers/myimage/Dockerfile
          tag_semver: true
          cache: true
```

**Comment by elgohr** [comment]
Do you build a Linux container?
Can you please provide the output?

**Comment by linkdd** [comment]
Yes it is a Linux container (based on alpine).

The output is pretty long because I use a multistage Dockerfile that rely heavily on the cache (locally a second build run in about 2s), so let me split it for you:

First, the ``docker pull`` to get the cache (but since I do not push intermediate containers, it pulls only the final stage):

```
2020-08-07T23:31:18.4233097Z Login Succeeded
2020-08-07T23:31:18.6192495Z latest: Pulling from myorga/myrepo/myimage
2020-08-07T23:31:18.6192989Z cbdbe7a5bc2a: Already exists
2020-08-07T23:31:18.6247779Z 13d9ba187739: Pulling fs layer
2020-08-07T23:31:18.6262998Z bec39fe62351: Pulling fs layer
2020-08-07T23:31:18.6263420Z f90cb0ea40ad: Pulling fs layer
2020-08-07T23:31:18.6263738Z 32b746c41de5: Pulling fs layer
2020-08-07T23:31:18.6264042Z fc83e958f48f: Pulling fs layer
2020-08-07T23:31:18.6264346Z 213bb4cd3f13: Pulling fs layer
2020-08-07T23:31:18.6264679Z 3e86107199c2: Pulling fs layer
2020-08-07T23:31:18.6265059Z bc693253d59b: Pulling fs layer
2020-08-07T23:31:18.6265337Z 4f4c816fde4d: Pulling fs layer
2020-08-07T23:31:18.6265937Z ae86814217ab: Pulling fs layer
2020-08-07T23:31:18.6266207Z 6cda2d78d926: Pulling fs layer
2020-08-07T23:31:18.6266481Z 32b746c41de5: Waiting
2020-08-07T23:31:18.6266751Z fc83e958f48f: Waiting
2020-08-07T23:31:18.6267035Z 213bb4cd3f13: Waiting
2020-08-07T23:31:18.6267373Z 3e86107199c2: Waiting
2020-08-07T23:31:18.6267635Z bc693253d59b: Waiting
2020-08-07T23:31:18.6267898Z 4f4c816fde4d: Waiting
2020-08-07T23:31:18.6268160Z ae86814217ab: Waiting
2020-08-07T23:31:18.6268425Z 6cda2d78d926: Waiting
2020-08-07T23:31:18.8893546Z f90cb0ea40ad: Verifying Checksum
2020-08-07T23:31:18.8894009Z f90cb0ea40ad: Download complete
2020-08-07T23:31:19.0506083Z bec39fe62351: Verifying Checksum
2020-08-07T23:31:19.0506408Z bec39fe62351: Download complete
2020-08-07T23:31:19.2142660Z 32b746c41de5: Verifying Checksum
2020-08-07T23:31:19.2144566Z 32b746c41de5: Download complete
2020-08-07T23:31:19.2727600Z fc83e958f48f: Verifying Checksum
2020-08-07T23:31:19.2727793Z fc83e958f48f: Download complete
2020-08-07T23:31:19.4948832Z 3e86107199c2: Download complete
2020-08-07T23:31:19.7292832Z bc693253d59b: Verifying Checksum
2020-08-07T23:31:19.7293178Z bc693253d59b: Download complete
2020-08-07T23:31:19.9523857Z 4f4c816fde4d: Verifying Checksum
2020-08-07T23:31:19.9524063Z 4f4c816fde4d: Download complete
2020-08-07T23:31:20.0161314Z 13d9ba187739: Verifying Checksum
2020-08-07T23:31:20.0161558Z 13d9ba187739: Download complete
2020-08-07T23:31:20.1633944Z 213bb4cd3f13: Verifying Checksum
2020-08-07T23:31:20.1634179Z 213bb4cd3f13: Download complete
2020-08-07T23:31:20.1774551Z ae86814217ab: Verifying Checksum
2020-08-07T23:31:20.1774771Z ae86814217ab: Download complete
2020-08-07T23:31:20.3275344Z 6cda2d78d926: Verifying Checksum
2020-08-07T23:31:20.3275516Z 6cda2d78d926: Download complete
2020-08-07T23:31:21.2031102Z 13d9ba187739: Pull complete
2020-08-07T23:31:21.2966200Z bec39fe62351: Pull complete
2020-08-07T23:31:21.3469198Z f90cb0ea40ad: Pull complete
2020-08-07T23:31:21.4169832Z 32b746c41de5: Pull complete
2020-08-07T23:31:21.4654139Z fc83e958f48f: Pull complete
2020-08-07T23:31:24.9482864Z 213bb4cd3f13: Pull complete
2020-08-07T23:31:25.0115345Z 3e86107199c2: Pull complete
2020-08-07T23:31:25.0607643Z bc693253d59b: Pull complete
2020-08-07T23:31:25.1110137Z 4f4c816fde4d: Pull complete
2020-08-07T23:31:25.1561994Z ae86814217ab: Pull complete
2020-08-07T23:31:25.2002627Z 6cda2d78d926: Pull complete
2020-08-07T23:31:25.2024679Z Digest: sha256:6ce69f4326b2d19daa10831d5a82580d332d51a3ca7ad93629089854d6458922
2020-08-07T23:31:25.2042437Z Status: Downloaded newer image for docker.pkg.github.com/myorga/myrepo/myimage:latest
2020-08-07T23:31:25.2083522Z docker.pkg.github.com/myorga/myrepo/myimage:latest
```

First stages, not using the cache (it does in local since it has):
```
2020-08-07T23:31:25.3134384Z Sending build context to Docker daemon  3.073MB
2020-08-07T23:31:25.3134548Z 
2020-08-07T23:31:25.3168582Z Step 1/31 : FROM scratch AS context
2020-08-07T23:31:25.3172126Z  ---> 
2020-08-07T23:31:25.3172292Z Step 2/31 : ADD sources/backend/package.json     sources/backend/yarn.lock     /workspace/
2020-08-07T23:31:25.4624034Z  ---> ba41c37c879e
2020-08-07T23:31:25.4624247Z Step 3/31 : FROM scratch AS sources
2020-08-07T23:31:25.4630562Z  ---> 
2020-08-07T23:31:25.4630783Z Step 4/31 : ADD sources/backend/main.js /workspace/main.js
2020-08-07T23:31:25.5536848Z  ---> f8c4f4eb086b
2020-08-07T23:31:25.5537119Z Step 5/31 : ADD sources/backend/src /workspace/src
2020-08-07T23:31:31.8420219Z  ---> da51676a782a
2020-08-07T23:31:31.8420578Z Step 6/31 : ADD sources/backend/prisma /workspace/prisma
2020-08-07T23:31:32.0103127Z  ---> c235f1283299
2020-08-07T23:31:32.0103399Z Step 7/31 : ADD sources/backend/.graphqlconfig.yml /workspace/.graphqlconfig.yml
2020-08-07T23:31:33.0069940Z  ---> 142decb7fb4a
2020-08-07T23:31:33.0070143Z Step 8/31 : FROM node:alpine AS dependencies
2020-08-07T23:31:33.1497929Z alpine: Pulling from library/node
2020-08-07T23:31:33.2191115Z cbdbe7a5bc2a: Already exists
2020-08-07T23:31:33.2214935Z 13d9ba187739: Already exists
2020-08-07T23:31:33.2242019Z bec39fe62351: Already exists
2020-08-07T23:31:33.2261352Z f90cb0ea40ad: Already exists
2020-08-07T23:31:33.3221338Z Digest: sha256:358a7a0ccdaa4bd4fbeca4a88f818bb96084ceb4db3a52b91622898e27c4c39b
2020-08-07T23:31:33.3241404Z Status: Downloaded newer image for node:alpine
2020-08-07T23:31:33.3255903Z  ---> 0d8a3475dbc3
2020-08-07T23:31:33.3256619Z Step 9/31 : COPY --from=context /workspace/package.json /workspace/package.json
2020-08-07T23:31:34.0211707Z  ---> 1f80322d5f8b
2020-08-07T23:31:34.0213101Z Step 10/31 : COPY --from=context /workspace/yarn.lock /workspace/yarn.lock
2020-08-07T23:31:35.0192445Z  ---> 1e5371b36371
2020-08-07T23:31:35.0192631Z Step 11/31 : WORKDIR /workspace
2020-08-07T23:31:35.0356686Z  ---> Running in c18157730d43
2020-08-07T23:31:36.0171295Z Removing intermediate container c18157730d43
2020-08-07T23:31:36.0171996Z  ---> eefc6b9bfbc1
2020-08-07T23:31:36.0172254Z Step 12/31 : RUN yarn install --network-timeout 300000 --production
2020-08-07T23:31:36.0326342Z  ---> Running in 603b51272b23
2020-08-07T23:31:36.4359051Z yarn install v1.22.4
2020-08-07T23:31:36.5044221Z [1/5] Validating package.json...
2020-08-07T23:31:36.5066208Z [2/5] Resolving packages...
2020-08-07T23:31:36.7996133Z [3/5] Fetching packages...
2020-08-07T23:31:50.2489178Z info fsevents@1.2.13: The platform "linux" is incompatible with this module.
2020-08-07T23:31:50.2491649Z info "fsevents@1.2.13" is an optional dependency and failed compatibility check. Excluding it from installation.
2020-08-07T23:31:50.2497329Z [4/5] Linking dependencies...
2020-08-07T23:31:50.2529682Z [91mwarning "graphql-yoga > graphql-playground-middleware-lambda@1.7.12" has unmet peer dependency "aws-lambda@^0.1.2".
2020-08-07T23:31:50.2532821Z [0m[91mwarning "graphql-yoga > graphql-subscriptions@0.5.8" has incorrect peer dependency "graphql@^0.10.5 || ^0.11.3 || ^0.12.0 || ^0.13.0".
2020-08-07T23:31:50.2537735Z [0m[91mwarning "graphql-yoga > apollo-server-express > apollo-server-core@1.4.0" has incorrect peer dependency "graphql@^0.9.0 || ^0.10.0 || ^0.11.0 || ^0.12.0 || ^0.13.0".
2020-08-07T23:31:50.2545574Z [0m[91mwarning "graphql-yoga > apollo-server-express > apollo-server-core > apollo-cache-control@0.1.1" has incorrect peer dependency "graphql@0.10.x - 0.13.x".
2020-08-07T23:31:50.2549192Z [0m[91mwarning "graphql-yoga > apollo-server-express > apollo-server-core > apollo-tracing@0.1.4" has incorrect peer dependency "graphql@0.10.x - 0.13.x".
2020-08-07T23:31:50.2551840Z [0m[91mwarning "graphql-yoga > apollo-server-express > apollo-server-core > graphql-extensions@0.0.10" has incorrect peer dependency "graphql@0.10.x - 0.13.x".
2020-08-07T23:31:55.2683160Z [0m[5/5] Building fresh packages...
2020-08-07T23:31:55.5167251Z Done in 19.09s.
2020-08-07T23:32:21.5889296Z Removing intermediate container 603b51272b23
2020-08-07T23:32:21.5890239Z  ---> cdd6f6407915
2020-08-07T23:32:21.5892072Z Step 13/31 : FROM node:alpine AS dev-dependencies
2020-08-07T23:32:21.5899206Z  ---> 0d8a3475dbc3
2020-08-07T23:32:21.5899657Z Step 14/31 : COPY --from=dependencies /workspace /workspace
2020-08-07T23:32:38.3408297Z  ---> 32f1a17a8877
2020-08-07T23:32:38.3408509Z Step 15/31 : WORKDIR /workspace
2020-08-07T23:32:38.3580772Z  ---> Running in c61a30c3fef8
2020-08-07T23:32:39.3056093Z Removing intermediate container c61a30c3fef8
2020-08-07T23:32:39.3056997Z  ---> 4d2340968b99
2020-08-07T23:32:39.3057423Z Step 16/31 : RUN yarn install --network-timeout 300000
2020-08-07T23:32:39.3366148Z  ---> Running in cad43751030c
2020-08-07T23:32:39.6832666Z yarn install v1.22.4
2020-08-07T23:32:39.7552760Z [1/5] Validating package.json...
2020-08-07T23:32:39.7578936Z [2/5] Resolving packages...
2020-08-07T23:32:40.0386621Z [3/5] Fetching packages...
2020-08-07T23:32:53.5571062Z info fsevents@1.2.13: The platform "linux" is incompatible with this module.
2020-08-07T23:32:53.5575969Z info "fsevents@1.2.13" is an optional dependency and failed compatibility check. Excluding it from installation.
2020-08-07T23:32:53.5585473Z [4/5] Linking dependencies...
2020-08-07T23:32:53.5618360Z [91mwarning "graphql-yoga > graphql-playground-middleware-lambda@1.7.12" has unmet peer dependency "aws-lambda@^0.1.2".
2020-08-07T23:32:53.5622242Z [0m[91mwarning "graphql-yoga > graphql-subscriptions@0.5.8" has incorrect peer dependency "graphql@^0.10.5 || ^0.11.3 || ^0.12.0 || ^0.13.0".
2020-08-07T23:32:53.5627320Z [0m[91mwarning "graphql-yoga > apollo-server-express > apollo-server-core@1.4.0" has incorrect peer dependency "graphql@^0.9.0 || ^0.10.0 || ^0.11.0 || ^0.12.0 || ^0.13.0".
2020-08-07T23:32:53.5635098Z [0m[91mwarning "graphql-yoga > apollo-server-express > apollo-server-core > apollo-cache-control@0.1.1" has incorrect peer dependency "graphql@0.10.x - 0.13.x".
2020-08-07T23:32:53.5639127Z [0m[91mwarning "graphql-yoga > apollo-server-express > apollo-server-core > apollo-tracing@0.1.4" has incorrect peer dependency "graphql@0.10.x - 0.13.x".
2020-08-07T23:32:53.5642814Z [0m[91mwarning "graphql-yoga > apollo-server-express > apollo-server-core > graphql-extensions@0.0.10" has incorrect peer dependency "graphql@0.10.x - 0.13.x".
2020-08-07T23:32:58.2404211Z [0m[5/5] Building fresh packages...
2020-08-07T23:32:58.8748217Z Done in 19.20s.
2020-08-07T23:33:25.4026532Z Removing intermediate container cad43751030c
2020-08-07T23:33:25.4027757Z  ---> 80cf33f00f0d
```

Final stage, using the cache:
```
2020-08-07T23:33:25.4028066Z Step 17/31 : FROM node:alpine AS runner
2020-08-07T23:33:25.4035041Z  ---> 0d8a3475dbc3
2020-08-07T23:33:25.4035527Z Step 18/31 : RUN apk add --no-cache bash
2020-08-07T23:33:25.6004977Z  ---> Using cache
2020-08-07T23:33:25.6005274Z  ---> c50583101975
2020-08-07T23:33:25.6005610Z Step 19/31 : COPY --from=context /workspace/package.json /workspace/package.json
2020-08-07T23:33:25.6057271Z  ---> Using cache
2020-08-07T23:33:25.7937543Z  ---> f3e6cfd0af2c
2020-08-07T23:33:25.7938158Z Step 20/31 : COPY --from=dependencies /workspace/node_modules /workspace/node_modules
2020-08-07T23:33:27.5883884Z  ---> Using cache
2020-08-07T23:33:27.6293255Z  ---> cf0d2847fabb
2020-08-07T23:33:27.6293879Z Step 21/31 : COPY --from=sources /workspace/src /workspace/src
2020-08-07T23:33:27.6368101Z  ---> Using cache
2020-08-07T23:33:27.6391407Z  ---> 47855615634e
2020-08-07T23:33:27.6391831Z Step 22/31 : COPY --from=sources /workspace/main.js /workspace/main.js
2020-08-07T23:33:27.6562491Z  ---> Using cache
2020-08-07T23:33:27.6562853Z  ---> b30e46ed3d8f
2020-08-07T23:33:27.6563337Z Step 23/31 : COPY --from=sources /workspace/.graphqlconfig.yml /workspace/.graphqlconfig.yml
2020-08-07T23:33:27.6563733Z  ---> Using cache
2020-08-07T23:33:27.6563959Z  ---> 6275a80c4280
2020-08-07T23:33:27.6564234Z Step 24/31 : COPY --from=sources /workspace/prisma /workspace/prisma
2020-08-07T23:33:27.6564467Z  ---> Using cache
2020-08-07T23:33:27.6586358Z  ---> 9bf9df1f863c
2020-08-07T23:33:27.6587112Z Step 25/31 : WORKDIR /workspace
2020-08-07T23:33:27.6607387Z  ---> Using cache
2020-08-07T23:33:27.6607647Z  ---> f7871aaf4d18
2020-08-07T23:33:27.6607989Z Step 26/31 : ADD containers/backend/entrypoint.sh     containers/backend/start.sh     containers/backend/wait-for.sh     /runtime/
2020-08-07T23:33:27.6638123Z  ---> Using cache
2020-08-07T23:33:27.6639058Z  ---> de573e9cd5b6
2020-08-07T23:33:27.6639348Z Step 27/31 : ENV PRISMA_ENDPOINT "http://localhost:4466/"
2020-08-07T23:33:27.6664857Z  ---> Using cache
2020-08-07T23:33:27.6665203Z  ---> ee879ccebba5
2020-08-07T23:33:27.6665363Z Step 28/31 : ENV NODE_ENV "production"
2020-08-07T23:33:27.6679018Z  ---> Using cache
2020-08-07T23:33:27.6679317Z  ---> 30142f2da99a
2020-08-07T23:33:27.6679484Z Step 29/31 : EXPOSE 4000
2020-08-07T23:33:27.6712467Z  ---> Using cache
2020-08-07T23:33:27.6712744Z  ---> d579fd5f3cd2
2020-08-07T23:33:27.6712902Z Step 30/31 : ENTRYPOINT ["/bin/sh", "/runtime/entrypoint.sh"]
2020-08-07T23:33:27.6737562Z  ---> Using cache
2020-08-07T23:33:27.6737839Z  ---> 71a154dfa39e
2020-08-07T23:33:27.6737990Z Step 31/31 : CMD ["/bin/sh", "/runtime/start.sh"]
2020-08-07T23:33:27.6746506Z  ---> Using cache
2020-08-07T23:33:27.6746782Z  ---> 963faa184856
2020-08-07T23:33:27.6862447Z Successfully built 963faa184856
2020-08-07T23:33:27.6880329Z Successfully tagged docker.pkg.github.com/myorga/myrepo/myimage:latest
```

So, to sums up the question, how do we push intermediate containers (not as separate images) so the ``docker pull`` can retrieve them?

**Comment by elgohr** [comment]
https://docs.semaphoreci.com/ci-cd-environment/docker-layer-caching/#the-add-command

**Comment by linkdd** [comment]
I don't understand your answer...

To know if the files have changed, Docker use their hash. From one build to another, if the hash doesn't change the cache is used.
You can see this in effect in step 19 to 26.

Here the problem is:

 - intermediate containers (stage that are not final) are removed and not pushed to the registry
 - on the next build, only the final stage is fetched (seems logical) so it does not have the cache for intermediate contaniers (therefore, every step trigger a cache-miss)

How do we push intermediate containers for the next build ?

**Comment by elgohr** [comment]
Ok, other way around. 
You start with a fresh worker, with no state. In order to compare a hash, you've got to have it, right?
As `ADD` adds an external resource, it has to build the container for that comparison.

**Comment by elgohr** [comment]
My bad, it should work if it's the same...
https://stackoverflow.com/a/43073309/1406669

**Comment by elgohr** [comment]
https://stackoverflow.com/a/43073580/1406669

**Comment by yusijs** [comment]
@elgohr I've started looking at this myself, and I'm not able to make this work. I can get it to work locally with buildkit by running
`DOCKER_BUILDKIT=1 docker build --cache-from=client:test .`
Which returns:
```
[+] Building 0.1s (23/23) FINISHED                                                                                                                                                                                                                                                                                                                                        
 => [internal] load .dockerignore                                                                                                                                                                                                                                                                                                                                    0.0s
 => => transferring context: 34B                                                                                                                                                                                                                                                                                                                                     0.0s
 => [internal] load build definition from Dockerfile                                                                                                                                                                                                                                                                                                                 0.0s
 => => transferring dockerfile: 37B                                                                                                                                                                                                                                                                                                                                  0.0s
 => [internal] load metadata for docker.io/library/nginx:alpine                                                                                                                                                                                                                                                                                                      0.0s
 => [internal] load metadata for docker.io/library/node:12-alpine                                                                                                                                                                                                                                                                                                    0.0s
 => importing cache manifest from client:test                                                                                                                                                                                                                                                                                                                        0.0s
 => [internal] load build context                                                                                                                                                                                                                                                                                                                                    0.1s
 => => transferring context: 31.75kB                                                                                                                                                                                                                                                                                                                                 0.1s
 => [stage-1 1/7] FROM docker.io/library/nginx:alpine                                                                                                                                                                                                                                                                                                                0.0s
 => [builder 1/9] FROM docker.io/library/node:12-alpine                                                                                                                                                                                                                                                                                                              0.0s
 => CACHED [stage-1 2/7] RUN apk add nodejs                                                                                                                                                                                                                                                                                                                          0.0s
 => CACHED [stage-1 3/7] COPY nginx-default.conf /etc/nginx/conf.d/default.conf                                                                                                                                                                                                                                                                                      0.0s
 => CACHED [stage-1 4/7] RUN rm -rf /usr/share/nginx/html/*                                                                                                                                                                                                                                                                                                          0.0s
 => CACHED [builder 2/9] COPY package.json yarn.lock ./                                                                                                                                                                                                                                                                                                              0.0s
 => CACHED [builder 4/9] RUN yarn                                                                                                                                                                                                                                                                                                                                    0.0s
 => CACHED [builder 5/9] RUN mkdir /client && cp -R ./node_modules ./client                                                                                                                                                                                                                                                                                          0.0s
 => CACHED [builder 6/9] WORKDIR /client                                                                                                                                                                                                                                                                                                                             0.0s
 => CACHED [builder 7/9] COPY src ./src                                                                                                                                                                                                                                                                                                                              0.0s
 => CACHED [builder 8/9] COPY .browserslistrc karma.conf.js *.json ./                                                                                                                                                                                                                                                                                                0.0s
 => CACHED [builder 9/9] RUN $(npm bin)/ng build --prod --build-optimizer --progress=false                                                                                                                                                                                                                                                                           0.0s
 => CACHED [stage-1 5/7] COPY --from=builder /client/dist/loss-control /usr/share/nginx/html                                                                                                                                                                                                                                                                         0.0s
 => CACHED [stage-1 6/7] COPY entrypoint.sh /entrypoint.sh                                                                                                                                                                                                                                                                                                           0.0s
 => CACHED [stage-1 7/7] COPY ./tools/environment-injector.js /environment-injector.js                                                                                                                                                                                                                                                                               0.0s
 => exporting to image                                                                                                                                                                                                                                                                                                                                               0.0s
 => => exporting layers                                                                                                                                                                                                                                                                                                                                              0.0s
 => => writing image sha256:ded428ec550f58aa7d7f75519e49ca87760cab86930d9082d569d534cf9c75ef   
```          

But using it in your action like this:
```
      - name: Publish client
        uses: elgohr/Publish-Docker-Github-Action@master
        env:
          DOCKER_BUILDKIT: 1
          BUILDKIT_INLINE_CACHE: 1
        with:
          name: ${{ steps.get_repo.outputs.REPO }}/client
          username: ${{ secrets.ACR_ID }}
          password: ${{ secrets.ACR_KEY }}
          cache: true
          buildargs: BUILDKIT_INLINE_CACHE,DOCKER_BUILDKIT
          tags: "${{steps.get_branch.outputs.BRANCH}}"
          workdir: client
```               
And only the latest stage is cached. 

It seems like perhaps the environment variables aren't preserved / used or something?

**Comment by elgohr** [comment]
Looks like it's not supported, yet.
https://github.community/t/github-actions-support-for-docker-buildkit/16083

**Comment by yusijs** [comment]
Latest comment says it is:

https://github.community/t/github-actions-support-for-docker-buildkit/16083/6

Also, buildkit works (based on the output). It seems more like the arguments are not passed, or environment variables are missing 🤔

**Comment by yusijs** [comment]
So I tried one of the images built from your action:
`DOCKER_BUILDKIT=1 docker build --cache-from=registry/repository/client:dev .`
And the output is as expected:
![image](https://user-images.githubusercontent.com/11261560/94030770-7b272d00-fdbe-11ea-9d6e-fb103d90a2d4.png)
Same command w/ DOCKER_BUILDKIT=0, and nothing is used from cache for the builder-stage

In your action however, this is the output:
![image](https://user-images.githubusercontent.com/11261560/94031098-cb05f400-fdbe-11ea-990d-df3cac01e7a9.png)

Without buildkit, it looks like this:
![image](https://user-images.githubusercontent.com/11261560/94031474-2cc65e00-fdbf-11ea-847b-c0afb82b0529.png)

I'm not really sure what's causing this, since it's obviously using buildkit (looking at my last 2 images), and I'm able to use the cache it generates when running the same command locally..

**Comment by elgohr** [comment]
The

**Comment by elgohr** [comment]
Looks like none of the Docker tools can check if a **remote** image exists.
There are solutions, like https://stackoverflow.com/a/39731444/1406669, which aren't compatible for every provider (Dockerhub, Github, ...).
In this way 🤷

**Comment by yusijs** [comment]
I dont think it breaks if the image is missing:
```
➜  client git:(dev) docker build --cache-from=hub.docker.com/repo/image:notexist .
Sending build context to Docker daemon  73.57MB
Step 1/18 : FROM node:12-alpine as builder
 ---> 057fa4cc38c2
Step 2/18 : ARG USE_PROXY
 ---> Running in d433bb33738c
Removing intermediate container d433bb33738c
 ---> a52cc44ae170
Step 3/18 : COPY package.json yarn.lock ./
 ---> 3ac050179090
```

So I'd say that safeguard is redundant, unless I'm missing something?

**Comment by elgohr** [comment]
It does, when you use buildkit:

```
DOCKER_BUILDKIT=1 docker build --cache-from utunbu:32r42 .
[+] Building 1.6s (7/7) FINISHED                                                
 => [internal] load .dockerignore                                          0.0s
 => => transferring context: 2B                                            0.0s
 => [internal] load build definition from Dockerfile                       0.0s
 => => transferring dockerfile: 69B                                        0.0s
 => [internal] load metadata for docker.io/library/ubuntu:latest           0.0s
 => ERROR importing cache manifest from utunbu:32r42                       1.4s
 => [1/2] FROM docker.io/library/ubuntu                                    0.0s
 => => resolve docker.io/library/ubuntu:latest                             0.0s
 => [2/2] RUN echo "test"                                                  0.2s
 => exporting to image                                                     0.0s
 => => exporting layers                                                    0.0s
 => => writing image sha256:8c6a89271f629f1739022f31d7ad48f1c60374a1f70a9  0.0s
------
 > importing cache manifest from utunbu:32r42:
------
```

`verifyImage` could be an idea.

**Comment by yusijs** [comment]
It doesnt seem like it failed to build ok though? It just said that it failed to find the image to build from cache, so it builds without one? I'll give it a spin later, or when I'm at work monday!
